pkgs/by-name/im/immich/package.nix

```diff
@@ -0,0 +1,318 @@
+        nativeBuildInputs = [
+          cacert
+          curl
+          unzip
+        ];
+
+        meta.license = lib.licenses.cc-by-40;
+      }
+      ''
+        mkdir $out
+        url="https://web.archive.org/web/20240724153050/http://download.geonames.org/export/dump"
+        curl -Lo ./cities500.zip "$url/cities500.zip"
+        curl -Lo $out/admin1CodesASCII.txt "$url/admin1CodesASCII.txt"
+        curl -Lo $out/admin2Codes.txt "$url/admin2Codes.txt"
+        curl -Lo $out/ne_10m_admin_0_countries.geojson \
+          https://raw.githubusercontent.com/nvkelso/natural-earth-vector/ca96624a56bd078437bca8184e78163e5039ad19/geojson/ne_10m_admin_0_countries.geojson
+
+        unzip ./cities500.zip -d $out/
+        echo "2024-07-24T15:30:50Z" > $out/geodata-date.txt
+      '';
+
+  src = fetchFromGitHub {
+    owner = "immich-app";
+    repo = "immich";
+    rev = "v${version}";
+    inherit (sources) hash;
+  };
+
+  openapi = buildNpmPackage' {
+    pname = "immich-openapi-sdk";
+    inherit version;
+    src = "${src}/open-api/typescript-sdk";
+    inherit (sources.components."open-api/typescript-sdk") npmDepsHash;
+
+    installPhase = ''
+      runHook preInstall
+
+      npm config delete cache
+      npm prune --omit=dev --omit=optional
+
+      mkdir -p $out
+      mv package.json package-lock.json node_modules build $out/
+
+      runHook postInstall
+    '';
+  };
+
+  cli = buildNpmPackage' {
+    pname = "immich-cli";
+    inherit version;
+    src = "${src}/cli";
+    inherit (sources.components.cli) npmDepsHash;
+
+    nativeBuildInputs = [ makeWrapper ];
+
+    preBuild = ''
+      rm node_modules/@immich/sdk
+      ln -s ${openapi} node_modules/@immich/sdk
+      # Rollup does not find the dependency otherwise
+      ln -s node_modules/@immich/sdk/node_modules/@oazapfts node_modules/
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      mkdir -p $out
+      mv package.json package-lock.json node_modules dist $out/
+
+      ls $out/dist
+
+      makeWrapper ${lib.getExe nodejs} $out/bin/immich --add-flags $out/dist/index.js
+
+      runHook postInstall
+    '';
+
+    meta = {
+      inherit (meta)
+        description
+        homepage
+        license
+        maintainers
+        platforms
+        ;
+      mainProgram = "immich";
+    };
+  };
+
+  web = buildNpmPackage' {
+    pname = "immich-web";
+    inherit version;
+    src = "${src}/web";
+    inherit (sources.components.web) npmDepsHash;
+
+    inherit (cli) preBuild;
+
+    installPhase = ''
+      runHook preInstall
+
+      cp -r build $out
+
+      runHook postInstall
+    '';
+  };
+
+  machine-learning = python3.pkgs.buildPythonApplication {
+    pname = "immich-machine-learning";
+    inherit version;
+    src = "${src}/machine-learning";
+    pyproject = true;
+
+    postPatch = ''
+      substituteInPlace pyproject.toml --replace-fail 'fastapi-slim' 'fastapi'
+
+      # Allow immich to use pydantic v2
```

Comment:
Why not use `packageOverrides` as in e.g. [pkgs/applications/video/frigate/default.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/video/frigate/default.nix)?

Reply:
because that causes build failures with other depdencies. This is far simpler

Reply:
You could at least file an upstream PR with this change.

Reply:
```diff
diff --git a/pkgs/by-name/im/immich/package.nix b/pkgs/by-name/im/immich/package.nix
index bed875eadd35..fa8c34a1754b 100644
--- a/pkgs/by-name/im/immich/package.nix
+++ b/pkgs/by-name/im/immich/package.nix
@@ -53,6 +53,17 @@ let
     packages = [ ];
   };
 
+  python = python3.override {
+    packageOverrides = self: super: {
+      pydantic = self.pydantic_1;
+
+      versioningit = super.versioningit.overridePythonAttrs (old: rec {
+        # incompatible with pydantic_1
+        doCheck = false;
+      });
+    };
+  };
+
   # The geodata website is not versioned, so we use the internet archive
   geodata =
     runCommand "immich-geodata"
@@ -163,7 +174,7 @@ let
     '';
   };
 
-  machine-learning = python3.pkgs.buildPythonApplication {
+  machine-learning = python.pkgs.buildPythonApplication {
     pname = "immich-machine-learning";
     inherit version;
     src = "${src}/machine-learning";
@@ -171,13 +182,6 @@ let
 
     postPatch = ''
       substituteInPlace pyproject.toml --replace-fail 'fastapi-slim' 'fastapi'
-
-      # Allow immich to use pydantic v2
-      substituteInPlace app/schemas.py --replace-fail 'pydantic' 'pydantic.v1'
-      substituteInPlace app/main.py --replace-fail 'pydantic' 'pydantic.v1'
-      substituteInPlace app/config.py \
-        --replace-fail 'pydantic' 'pydantic.v1' \
-        --replace-fail '/cache' '/var/cache/immich'
     '';
 
     pythonRelaxDeps = [
@@ -186,13 +190,13 @@ let
     ];
     pythonRemoveDeps = [ "opencv-python-headless" ];
 
-    build-system = with python3.pkgs; [
+    build-system = with python.pkgs; [
       poetry-core
       cython
     ];
 
     dependencies =
-      with python3.pkgs;
+      with python.pkgs;
       [
         insightface
         opencv4
@@ -218,10 +222,10 @@ let
       mkdir -p $out/share
       cp log_conf.json $out/share
 
-      cp -r ann $out/${python3.sitePackages}/
+      cp -r ann $out/${python.sitePackages}/
 
-      makeWrapper ${lib.getExe python3.pkgs.gunicorn} $out/bin/machine-learning \
-        --prefix PYTHONPATH : "$out/${python3.sitePackages}:$PYTHONPATH" \
+      makeWrapper ${lib.getExe python.pkgs.gunicorn} $out/bin/machine-learning \
+        --prefix PYTHONPATH : "$out/${python.sitePackages}:$PYTHONPATH" \
         --set-default MACHINE_LEARNING_WORKERS 1 \
         --set-default MACHINE_LEARNING_WORKER_TIMEOUT 120 \
         --set-default IMMICH_HOST 127.0.0.1 \
@@ -268,7 +272,7 @@ buildNpmPackage' {
 
   nativeBuildInputs = [
     pkg-config
-    python3
+    python
     makeWrapper
     glib.dev
     nodePackages.node-gyp

```

Reply:
I am testing that but it basically rebuilds the whole python world. Is that really a good use of hydra resources?

Reply:
> Is that really a good use of hydra resources?

No. I'm gonna file an upstream PR so we can use pydantic v2.

Reply:
You can use `fetchpatch2` now.

Reply:
ensure to use `?full_index=1` at the end of the URL